java/ql/lib/semmle/code/java/regex/regex.qll

```diff
@@ -0,0 +1,933 @@
+import java
+private import RegexFlowConfigs
+
+// In all ranges handled by this library, `start` is inclusive and `end` is exclusive.
+/**
+ * A string literal that is used as a regular expression.
+ */
+abstract class RegexString extends StringLiteral {
+  /** Holds if a character set starts between `start` and `end`, including any negation character (`^`). */
+  private predicate charSetStart0(int start, int end) {
+    this.nonEscapedCharAt(start) = "[" and
+    (if this.getChar(start + 1) = "^" then end = start + 2 else end = start + 1)
+  }
+
+  /** Holds if the character at `pos` marks the end of a character class. */
+  private predicate charSetEnd0(int pos) {
+    this.nonEscapedCharAt(pos) = "]" and
+    /* special case: `[]]` and `[^]]` are valid char classes.  */
```

Comment:
Are both `[^]` and `[^]]` valid char classes? Are we handling both correctly?

Reply:
`[^]` isn't a valid class.

Reply:
`InputSymbol.toString` in ReDoSUtil.qll indicates that it is, but that might be language-dependent? I would hope that a thing like `[^]` had mostly engine-independent semantics, but that might be too much to hope for?

Reply:
It's a valid class in JS, in some other dialects it's not.  
(I'm not sure about Java)

Reply:
It is in some regex dialects, but not in Java.